luda-editor/client/src/pages/graphic_asset_manage_page/cg_viewer/mod.rs

```diff
@@ -0,0 +1,269 @@
+mod part_picker;
+
+use self::part_picker::PartPicker;
+use crate::{color, components::cg_render::CgRender, storage::get_project_cg_thumbnail_image_url};
+use namui::prelude::*;
+use namui_prebuilt::{button, simple_rect, table::hooks::*, typography};
+use rpc::data::{ScreenCg, ScreenCgPart};
+use std::ops::Deref;
+
+const MODAL_MAX_WH: Wh<Px> = Wh {
+    width: px(1280.0),
+    height: px(720.0),
+};
+const MODAL_MIN_MARGIN: Px = px(16.0);
+const TITLE_BAR_HEIGHT: Px = px(48.0);
+const PICKER_WIDTH: Px = px(512.0);
+
+#[component]
+pub struct CgViewer<'a> {
+    pub wh: Wh<Px>,
+    pub cg_file: &'a rpc::data::CgFile,
+    pub project_id: Uuid,
+    pub on_close: &'a dyn Fn(),
+}
+
+impl Component for CgViewer<'_> {
+    fn render(self, ctx: &RenderCtx) -> RenderDone {
+        let Self {
+            wh,
+            cg_file,
+            project_id,
+            on_close,
+        } = self;
+
+        let wh = ctx.track_eq(&wh);
+        let modal_rect = ctx.memo(|| {
+            let modal_wh = Wh {
+                width: (wh.width - MODAL_MIN_MARGIN * 2.0).min(MODAL_MAX_WH.width),
+                height: (wh.height - MODAL_MIN_MARGIN * 2.0).min(MODAL_MAX_WH.height),
+            };
+            let modal_xy = Xy {
+                x: (wh.width - modal_wh.width) / 2.0,
+                y: (wh.height - modal_wh.height) / 2.0,
+            };
+            Rect::from_xy_wh(modal_xy, modal_wh)
+        });
```

Comment:
This will make a problem when wh is changed. I guess you don't have to memorize this value because it's very simple calculation.

Reply:
Fixed with b47d351